### PR TITLE
RESTWS-640 Implement search by partial identifier

### PR DESCRIPTION
**Pull request by AdamGrzybkowski**

### CODE REVIEW

File: omod-1.10/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_10/DrugController1_10Test.java [hunk 1]
```diff
@@ -102,7 +102,7 @@ public void getDrugsByMapping_shouldReturnADrugThatMatchesTheCodeAndSourceAndThe
 		results = deserialize(handle(newGetRequest(getURI(), new Parameter("s", "getDrugsByMapping"), new Parameter("code",
 		        "CD41003"), new Parameter("source", sourceUuid), new Parameter("preferredMapTypes", mapTypeUuids))));
 		
-		assertEquals(expectedDrugUuid, PropertyUtils.getProperty(Util.getResultsList(results).get(0), "uuid"));
+		assertEquals(expectedDrugUuid, PropertyUtils.getProperty(Util.getResultsList(results).get(1), "uuid"));
```
Comment (AdamGrzybkowski) [inline omod-1.10/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_10/DrugController1_10Test.java]: This test started to fail... edited this only to show which one but nor sure why this started to fail...

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/api/RestHelperService.java [hunk 2]
```diff
@@ -31,6 +32,9 @@
 	
 	List<Patient> getPatients(Collection<Integer> patientIds);
 	
+	@Authorized({ "View Patients" })
+	List<Patient> findPatients(String identifier, String searchType, boolean includeVoided);
```
Comment (rkorytkowski) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/api/RestHelperService.java]: Rename to findPatientsByIdentifier.

Comment (AdamGrzybkowski) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/api/RestHelperService.java]: @rkorytkowski Done, but please take a look at the above test. Any idea why it stared to fail ?

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/api/impl/RestHelperServiceImpl.java [hunk 3]
```diff
@@ -125,6 +126,16 @@ private Session getSession() {
 		return ret;
 	}
 	
+	@Override
+	public List<Patient> findPatientsByIdentifierStartingWith(String identifier, boolean includeVoided) {
+		Criteria criteria = getSession().createCriteria(Patient.class);
+		criteria.createAlias("identifiers", "identifiers");
+		criteria.add(Restrictions.like("identifiers.identifier", identifier, MatchMode.START));
+		criteria.setResultTransformer(Criteria.DISTINCT_ROOT_ENTITY);
+		criteria.add(Restrictions.eq("voided", includeVoided));
```
Comment (rkorytkowski) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/api/impl/RestHelperServiceImpl.java]: As it is now if you call it with includeVoided=true only voided patients will be returned, whereas both voided and non voided patients should be returned.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java [hunk 4]
```diff
@@ -43,14 +48,23 @@ public SearchConfig getSearchConfig() {
...
 		
 		String identifier = context.getRequest().getParameter("identifier");
+		String searchType = context.getParameter("searchType");
 		
 		if (StringUtils.isNotBlank(identifier)) {
-			List<Patient> patients = Context.getPatientService().getPatients(null, identifier, null, true);
-			if (patients != null && patients.size() > 0) {
-				return new NeedsPaging<Patient>(patients, context);
+			if("start".equals(searchType)){
+				List<Patient> patients = restHelperService.findPatientsByIdentifierStartingWith(identifier, false);
```
Comment (rkorytkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: Is it okay to always return only non voided patients? Please add support for includeAll=true as per https://wiki.openmrs.org/display/docs/REST+Web+Services+API+For+Clients

Call context.getIncludeAll()

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java [hunk 5]
```diff
@@ -43,14 +48,23 @@ public SearchConfig getSearchConfig() {
...
+				if (patients != null && patients.size() > 0) {
+					return new NeedsPaging<Patient>(patients, context);
+				}
+			} else {
+				List<Patient> patients = Context.getPatientService().getPatients(null, identifier, null, true);
```
Comment (rkorytkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: Could you handle getIncludeAll here as well?

Comment (AdamGrzybkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: @rkorytkowski This needs to be done in openmrs core I think. The method used here does not have this argument.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java (hunk 4 above)
Comment (AdamGrzybkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: There is no such method

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/api/impl/RestHelperServiceImpl.java (hunk 3 above)
Comment (AdamGrzybkowski) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/api/impl/RestHelperServiceImpl.java]: done

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java (hunk 5 above)
Comment (rkorytkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: I see fair enough... Let's leave it as is here since it was that before and is out of scope for this issue.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java [hunk 6]
```diff
@@ -43,14 +48,23 @@ public SearchConfig getSearchConfig() {
...
 		
 		String identifier = context.getRequest().getParameter("identifier");
+		String searchType = context.getParameter("searchType");
 		
 		if (StringUtils.isNotBlank(identifier)) {
-			List<Patient> patients = Context.getPatientService().getPatients(null, identifier, null, true);
-			if (patients != null && patients.size() > 0) {
-				return new NeedsPaging<Patient>(patients, context);
+			if ("start".equals(searchType)) {
+				List<Patient> patients = restHelperService.findPatientsByIdentifierStartingWith(identifier, false);
```
Comment (rkorytkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: Still no support for includeAll here? You are always passing false...

Comment (AdamGrzybkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: I did write comment above that there is no such method as Context.getIncludeAll()

Comment (rkorytkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: it's context.getIncludeAll() not Context.getIncludeAll() (method parameter of type RequestContext)

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java (hunk 5 above)
Comment (AdamGrzybkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: ... oh good. Sorry for that... Fixed

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java (hunk 6 above)
Comment (AdamGrzybkowski) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/PatientByIdentifierSearchHandler1_8.java]: ... oh good. Sorry for that... Fixed
